### PR TITLE
[WINSPOOL] EnumPrintersA(): Fix failure handling on/after EnumPrintersW(). CORE-16622

### PR DESCRIPTION
**Pull request by SergeGautherie**
Split off from #2298.

Addendum to e8b1778.
JIRA issue: [CORE-16622](https://jira.reactos.org/browse/CORE-16622)

### CODE REVIEW

**Comment by SergeGautherie** [comment]
Cc @ColinFinck

File: win32ss/printing/base/winspool/printers.c [hunk 1]
```diff
@@ -449,9 +449,17 @@ EnumPrintersA(DWORD Flags, PSTR Name, DWORD Level, PBYTE pPrinterEnum, DWORD cbB
  
     /* Ref: https://stackoverflow.com/questions/41147180/why-enumprintersa-and-enumprintersw-request-the-same-amount-of-memory */
     bReturnValue = EnumPrintersW(Flags, pwszName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
-    HeapFree(hProcessHeap, 0, pwszName);
+    TRACE("*pcReturned is '%d' and bReturnValue is '%d' and GetLastError is '%ld'.\n",
+          *pcReturned, bReturnValue, GetLastError());
 
-    TRACE("*pcReturned is '%d' and bReturnValue is '%d' and GetLastError is '%ld'.\n", *pcReturned, bReturnValue, GetLastError());
+    if (pwszName)
+        HeapFree(hProcessHeap, 0, pwszName);
```
Comment (ThFabba) [inline win32ss/printing/base/winspool/printers.c]: Add braces please.
This is okay since you're changing something else right next to it (or in new code), but for future reference, HeapFree accepts null pointers and it's not necessary to add null checks to existing code.

File: win32ss/printing/base/winspool/printers.c [hunk 2]
```diff
@@ -449,9 +449,17 @@ EnumPrintersA(DWORD Flags, PSTR Name, DWORD Level, PBYTE pPrinterEnum, DWORD cbB
  
     /* Ref: https://stackoverflow.com/questions/41147180/why-enumprintersa-and-enumprintersw-request-the-same-amount-of-memory */
     bReturnValue = EnumPrintersW(Flags, pwszName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
-    HeapFree(hProcessHeap, 0, pwszName);
+    TRACE("*pcReturned is '%d' and bReturnValue is '%d' and GetLastError is '%ld'.\n",
+          *pcReturned, bReturnValue, GetLastError());
 
-    TRACE("*pcReturned is '%d' and bReturnValue is '%d' and GetLastError is '%ld'.\n", *pcReturned, bReturnValue, GetLastError());
```
Comment (ThFabba) [inline win32ss/printing/base/winspool/printers.c]: Not necessary.

File: win32ss/printing/base/winspool/printers.c [hunk 3]
```diff
@@ -449,9 +449,17 @@ EnumPrintersA(DWORD Flags, PSTR Name, DWORD Level, PBYTE pPrinterEnum, DWORD cbB
...
+    if (pwszName)
+        HeapFree(hProcessHeap, 0, pwszName);
+
+    if (!bReturnValue)
```
Comment (ThFabba) [inline win32ss/printing/base/winspool/printers.c]: Good catch, thanks!

**Comment by SergeGautherie** [comment]
> GetLastError()

Well, MSDN says HeapFree() sets LastError. Is that false?
(And the line is a bit long...)

> HeapFree()

Done. I know. I simply tried to follow this file style...

**Comment by ThFabba** [comment]
> > GetLastError()
> 
> Well, MSDN says HeapFree() sets LastError. Is that false?
> (And the line is a bit long...)

I'm only objecting to the style change, not the reordering.
It's really hard to follow conversations this way. Actually replying to code/comments would be more useful.

**Comment by SergeGautherie** [comment]
> style change

Reverted.

> replying to code/comments

I know. Sadly, many GitHub features do not work on my usual browser since two weeks ago.

File: win32ss/printing/base/winspool/printers.c (hunk 3 above)
Comment (Doug-Lyons) [reply win32ss/printing/base/winspool/printers.c]: This call is expected to fail with a GetLastError of ERROR_INSUFFICIENT_BUFFER when cbBuf = 0 when we probe to see how large the buffer should be. Should this condition be checked and the failure message be suppressed in this case?

**Comment by Doug-Lyons** [comment]
This call to EnumPrintersW is expected to fail with a GetLastError of ERROR_INSUFFICIENT_BUFFER when cbBuf = 0 when we probe to see how large the buffer should be. Should this condition be checked and the new failure message be suppressed in this case? Otherwise, this output is not very helpful.

**Comment by SergeGautherie** [comment]
> new failure message be suppressed

I agree: it should simply be removed actually.

To be fixed further, soon.

File: win32ss/printing/base/winspool/printers.c (hunk 1 above)
Comment (SergeGautherie) [reply win32ss/printing/base/winspool/printers.c]: > HeapFree()

Done. I know. I simply tried to follow this file style...

File: win32ss/printing/base/winspool/printers.c (hunk 2 above)
Comment (SergeGautherie) [reply win32ss/printing/base/winspool/printers.c]: > GetLastError()

Well, MSDN says HeapFree() sets LastError. Is that false?
(And the line is a bit long...)

File: win32ss/printing/base/winspool/printers.c (hunk 3 above)
Comment (SergeGautherie) [reply win32ss/printing/base/winspool/printers.c]: > failure message be suppressed

I agree: it should simply be removed actually.

File: win32ss/printing/base/winspool/printers.c [hunk 4]
```diff
@@ -448,10 +449,20 @@ EnumPrintersA(DWORD Flags, PSTR Name, DWORD Level, PBYTE pPrinterEnum, DWORD cbB
...
     /* Ref: https://stackoverflow.com/questions/41147180/why-enumprintersa-and-enumprintersw-request-the-same-amount-of-memory */
-    bReturnValue = EnumPrintersW(Flags, pwszName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
-    HeapFree(hProcessHeap, 0, pwszName);
+    bResult = EnumPrintersW(Flags, pwszName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
+    dwErrorCode = GetLastError();
 
-    TRACE("*pcReturned is '%d' and bReturnValue is '%d' and GetLastError is '%ld'.\n", *pcReturned, bReturnValue, GetLastError());
+    TRACE("*pcReturned is '%d' and bResult is '%d' and dwErrorCode is '%ld'.\n", *pcReturned, bResult, dwErrorCode);
+
+    if (pwszName)
```
Comment (ColinFinck) [inline win32ss/printing/base/winspool/printers.c]: Now that we have the recurring `if (pointer) HeapFree(pointer)` pattern in this function, please move it into the `Cleanup` branch to stay consistent with the rest of the printing code.

File: win32ss/printing/base/winspool/printers.c [hunk 5]
```diff
@@ -448,10 +449,20 @@ EnumPrintersA(DWORD Flags, PSTR Name, DWORD Level, PBYTE pPrinterEnum, DWORD cbB
     }
  
     /* Ref: https://stackoverflow.com/questions/41147180/why-enumprintersa-and-enumprintersw-request-the-same-amount-of-memory */
-    bReturnValue = EnumPrintersW(Flags, pwszName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
-    HeapFree(hProcessHeap, 0, pwszName);
+    bResult = EnumPrintersW(Flags, pwszName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
+    dwErrorCode = GetLastError();
```
Comment (ColinFinck) [inline win32ss/printing/base/winspool/printers.c]: This is only relevant for the failure case, so please move it into the `if (!bResult)` condition. Follows the usual `if (!returnvalue) GetLastError()` pattern.

File: win32ss/printing/base/winspool/printers.c (hunk 4 above)
Comment (SergeGautherie) [reply win32ss/printing/base/winspool/printers.c]: I concur. Done.

File: win32ss/printing/base/winspool/printers.c (hunk 5 above)
Comment (SergeGautherie) [reply win32ss/printing/base/winspool/printers.c]: I mostly concur and would do it if `TRACE()` was (re)moved too.
With `HeapFree()` previously, still with `TRACE()` and possibly with future code in between, I think it is safer to do it like this.

**Comment by JoachimHenze** [comment]
SergeGautherie, will this PR fix https://jira.reactos.org/browse/CORE-16715 ?

**Comment by SergeGautherie** [comment]
> will this PR fix [CORE-16715](https://jira.reactos.org/browse/CORE-16715)?

I don't know. I am simply working on CORE-16622 code itself.

File: win32ss/printing/base/winspool/printers.c (hunk 5 above)
Comment (SergeGautherie) [reply win32ss/printing/base/winspool/printers.c]: Eventually, I added 2 commits, so you can choose from.
